src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Helpers.cs

```diff
@@ -41,15 +41,22 @@ public static partial class JsonSerializer
 
         private static void WriteUsingMetadata<TValue>(Utf8JsonWriter writer, in TValue value, JsonTypeInfo jsonTypeInfo)
         {
-            WriteStack state = default;
-            state.Initialize(jsonTypeInfo, supportContinuation: false);
+            if (jsonTypeInfo is JsonTypeInfo<TValue> typedInfo && typedInfo.UseFastPathOnWrite)
+            {
+                typedInfo.Serialize!(writer, value);
```

Comment:
Can we use a `MemberNotNullWhen` attribute on `UseFastPathOnWrite` so this doesn't need to use a `!`?

Reply:
That would have been really nice but I moved the check to a different type per https://github.com/dotnet/runtime/pull/53212#discussion_r639317525.